Fix parse PIRSF

laisecf opened the pull request:
Fixes:
- fixing the field (sequenceLength and some positions) to calculate ratios (ovl and r) 🟢 
- the score to calculate ratios is the match one instead the locations one 🟢 
- at end we just keep the best match and its subfamilies 🟢 

We had a significant difference comparing PIRSF results on i5 and i6:
```
Matches only in expected : 1667
Matches only in observed : 381
Matches in both          : 2821
```

After fix still remains some differences (regression test using human proteome) - reason: slightly different scores.
```
Matches only in expected : 4
Matches only in observed : 7
Matches in both          : 4476
```

- merging domains for the same accession 🟢 
this fix will be made in i5 side too soon

HobnobMancer commented:
> I am working on the complete fix to beta version
Are you still working on this fix or is it complete and ready to review?

HobnobMancer reviewed (changes requested):
HmmBounds are always null:
```json
          "locations" : [
            {
              "start" : 2,
              "end" : 277,
              "representative" : true,
              "hmmStart" : 22,
              "hmmEnd" : 269,
              "hmmLength" : 431,
              "hmmBounds" : null,
              "evalue" : 1.9E-31,
              "score" : 105.19999694824219,
              "envelopeStart" : 22,
              "envelopeEnd" : 237,
              "location-fragments" : [
                {
                  "start" : 2,
                  "end" : 277,
                  "dcStatus" : "CONTINUOUS"
                }
              ]
            },
```
The envelope is always wrong. I ran the same HMMER3 dtbl through `PARSE_PIRSF` and the PIRSF perl script and the envelope start and end differs for every match, and consequently this carried over into comparing the i5 and i6 matches.

laisecf commented:
> HmmBounds are always null:
> 
> ```json
>           "locations" : [
>             {
>               "start" : 2,
>               "end" : 277,
>               "representative" : true,
>               "hmmStart" : 22,
>               "hmmEnd" : 269,
>               "hmmLength" : 431,
>               "hmmBounds" : null,
>               "evalue" : 1.9E-31,
>               "score" : 105.19999694824219,
>               "envelopeStart" : 22,
>               "envelopeEnd" : 237,
>               "location-fragments" : [
>                 {
>                   "start" : 2,
>                   "end" : 277,
>                   "dcStatus" : "CONTINUOUS"
>                 }
>               ]
>             },
> ```
> 
> The envelope is always wrong. I ran the same HMMER3 dtbl through `PARSE_PIRSF` and the PIRSF perl script and the envelope start and end differs for every match, and consequently this carried over into comparing the i5 and i6 matches.
> 
> There are also discrepancies between the i6 output and the output generated by the PIRSF Perl script, even though both apply post-processing to identical HMMER3 results - unlike in i5 where mismatches might arise from differences in HMMER3 matches themselves, as i5 does not use a Z-value.
> 
> ```shell
> < 0008FBEF8899F159525153FD741BE931      PIRSF036643     5       747
> < 0008FBEF8899F159525153FD741BE931      PIRSF036643     874     926
> < A37CD721605DEB2218223EBC024EFCB8      PIRSF005813     1       876
> > A37CD721605DEB2218223EBC024EFCB8      PIRSF037677     7       858
> < C84B0963DF32B12896E7154FF7B073AD      PIRSF005813     1       884
> > C84B0963DF32B12896E7154FF7B073AD      PIRSF037677     7       848
> < CDCA49D054C0AE4C50EBCB67211DD20C      PIRSF005813     1       877
> > CDCA49D054C0AE4C50EBCB67211DD20C      PIRSF037677     7       853
> > D519491AC20B1682AA63D72B573677E3      PIRSF001220     1       314
> < D8C5B6F9F0C715CD63B25FEB0F23EAE9      PIRSF029394     2       131
> ============ Summary ============
> Matches only in expected : 6
> Matches only in observed : 4
> Matches in both          : 166
> ```

Thank you!

About the construction of rawMatches:
There are some differences compared to i5 (as seq lenght), but the location positions remain the same. Here's an example I found while debugging:
i5 [row](https://github.com/ebi-pf-team/interproscan/blob/27c44e0c54984e6021e04f3212699098613f07c7/core/jms-implementation/support-mini-x86-32/bin/pirsf/3.10/PIRSF.pm#L207) (using print "" . join(", ", @$row) . "\n";):
Nematode_astacin, PIRSF036365, 612, 1, -, 485, 0.0017, 14.9, 0.2, 2, 2, 4.6e-05, 0.038, 10.4, 0.4, 225, 239, 186, 200, 181, 203, 0.91, -

i6 [row](https://github.com/ebi-pf-team/interproscan6/blob/main/modules/pirsf/main.nf#L50) (using println(lineData)):
[FF4BE3A65A9DE1B7DF61EAB375C49448, -, 485, Nematode_astacin, PIRSF036365, 612, 5.1e-07, 14.9, 0.2, 2, 2, 1.1e-05, 1.1e-05, 10.4, 0.4, 225, 239, 186, 200, 181, 203, 0.91, -]

About the envelope and hmmBounds:
Now that you comment about the wrong envelope positions I understood why some positions were different on script: For PIRSF the start/end positions in the output refer to the envelope positions. 
So I kept the original positions to calculate the ratios, since that was the issue I fixed. 
I also put back the function that creates the locations and HmmBounds. So, at this moment, we updated the start and end values using envelopeStart and envelopeEnd (to not affect the ratios). With this change now we have (for human proteome):
```
Matches only in expected : 88
Matches only in observed : 66
Matches in both          : 4400
```

Investigating those remains mismatches but this observations already help me to understand better!! o/

HobnobMancer reviewed (changes requested):
There are cases where the i6 `PARSE_PIRSF` module picks a different model to the perl script when both are analysing the same [HMMER3 domtbl file](https://github.com/user-attachments/files/20245555/pirsf.zip).
Here are some examples:
```bash
< A37CD721605DEB2218223EBC024EFCB8      PIRSF005813     1       876
> A37CD721605DEB2218223EBC024EFCB8      PIRSF037677     7       858
< C84B0963DF32B12896E7154FF7B073AD      PIRSF005813     1       884
> C84B0963DF32B12896E7154FF7B073AD      PIRSF037677     7       848
< CDCA49D054C0AE4C50EBCB67211DD20C      PIRSF005813     1       877
> CDCA49D054C0AE4C50EBCB67211DD20C      PIRSF037677     7       853
> D519491AC20B1682AA63D72B573677E3      PIRSF001220     1       314
< D8C5B6F9F0C715CD63B25FEB0F23EAE9      PIRSF029394     2       131
```

This is because of this change: " 🟡  using evalue ASC, score DESC to sort the best match (standardising with other members)". I was talking with Thiago to understand why PIRSF uses just the score to choose the best match and he explained that he just kept it as it was in the oldest code, but that it makes sense to update to evalue ASC, score DESC, like the other members. Another difference: i5 uses the score of the match (not of the location) in this part, I am in doubt about what is the most appropriate, mainly after the fix we will make in i5/i6 PIRSF to report a unique match (the one Matthias explain on Friday). Let's discuss this further tomorrow at our meeting o/

laisecf commented:
> ```shell
> D8C5B6F9F0C715CD63B25FEB0F23EAE9
> ```

Can you send me the fasta file related to this mismatches just to check if the results are the same using just the score? (the fasta in zip file is the symbolic link)

HobnobMancer commented:
> > ```shell
> > D8C5B6F9F0C715CD63B25FEB0F23EAE9
> > ```
> 
> Can you send me the fasta file related to those mismatches just to check if the results are the same using just the score? (the fasta in zip file is the symbolic link)

It's the `test.faa` file

laisecf commented:
> > > ```shell
> > > D8C5B6F9F0C715CD63B25FEB0F23EAE9
> > > ```
> > 
> > 
> > Can you send me the fasta file related to those mismatches just to check if the results are the same using just the score? (the fasta in zip file is the symbolic link)
> 
> It's the `test.faa` file
> 
> > This is because of this change: " 🟡 using evalue ASC, score DESC to sort the best match (standardising with other members)". I was talking with Thiago to understand why PIRSF uses just the score to choose the best match and he explained that he just kept it as it was in the oldest code, but that it makes sense to update to evalue ASC, score DESC, like the other members. Another difference: i5 uses the score of the match (not of the location) in this part, I am in doubt about what is the most appropriate, mainly after the fix we will make in i5/i6 PIRSF to report a unique match (the one Matthias explain on Friday). Let's discuss this further tomorrow at our meeting o/
> 
> The Perl script would have used the score because the E-value would be unreliable due a Z-value not being used (remember that the E-value is dependent on the database size), so it is a very logical decision to be implemented in i5. You could only implement screening with an E-value and score if you can guarantee a reproducible E-value so doing that before implementing a Z-value would have made no sense.

Make sense!! Ok, I will put back the oldest way (just using the score to best match), if we decide to put the z-score for i6 we can change to evalue ASC, score DESC ^^

HobnobMancer reviewed (changes requested):
I'm consistently finding differences when comparing the output of `PARSE_PIRSF` and the perl script for various [proteomes](https://github.com/user-attachments/files/20300958/domtbl_files.zip). It's a small number by `PARSE_PIRSF` is a groovy implementation of the perl module so we shouldn't be seeing any differences when the input is the same.

Testing using a _Phytophthora megakarya_ proteome:
```bash
< 1224CA081163A9B391E6EFE4CB065391      PIRSF000077     1       104
> 294A405EB457227AC5E88CA9DFD519B8      PIRSF001220     62      406
> 88C144D91676159D802B9271FB2464D8      PIRSF019663     9       238
> BC3C9B7C2BFD53A9C3706A04F5B4F6EC      PIRSF000778     1       132
< BE495BC3E429D23BB6DC48DC8C65CFB1      PIRSF000654     7       263
> FB5A78AE553D9E05A80C77A9268D957D      PIRSF000904     7       333
============ Summary ============
Matches only in expected : 2
Matches only in observed : 4
Matches in both          : 627
```
Testing using a _Fusarium oxysporum_ proteome:
```bash
> 0917AEB0B21ED38129863686AD88BD43      PIRSF000778     1       137
> 2B88E803A198E201C4772AFBC3E37A60      PIRSF001220     18      73
> 2B88E803A198E201C4772AFBC3E37A60      PIRSF001220     89      386
> 34C7DCFF982954C7FE7ABB61956B6382      PIRSF000124     14      91
> 34C7DCFF982954C7FE7ABB61956B6382      PIRSF000124     501     601
> 34C7DCFF982954C7FE7ABB61956B6382      PIRSF000124     97      477
> 3FAD2A5DA39E7C0B6E20A4748BE1B77F      PIRSF000778     14      154
> 4C1189E51A8C227FB859BB7743596AE7      PIRSF000124     76      451
> 7018EE411546F8029C4130D023B22C17      PIRSF019663     4       329
< 7ABD13AC026ACDB0EC99A1684C8B4A85      PIRSF036603     463     560
< 7ABD13AC026ACDB0EC99A1684C8B4A85      PIRSF036603     9       274
> A23F4E183B1D1B3CC7B8FF3B2446657F      PIRSF001220     1       225
> BCC5562BE28E9028E0CE54B5C212A631      PIRSF000124     14      28
> BCC5562BE28E9028E0CE54B5C212A631      PIRSF000124     158     538
> BCC5562BE28E9028E0CE54B5C212A631      PIRSF000124     562     662
> BCC5562BE28E9028E0CE54B5C212A631      PIRSF000124     82      152
> C2542CC2AB3292AD8DD57C01A4A48858      PIRSF000124     49      429
> CE414C5462C4094CACA00A3001D45B81      PIRSF000124     50      423
> E6B761F0C53F73524F8926B1BAAFDF61      PIRSF000124     63      432
> EB8C6264D585C439A907B7CE64D3A9FC      PIRSF000904     1       250
============ Summary ============
Matches only in expected : 2
Matches only in observed : 18
Matches in both          : 1248
```
Testing using a _Pectobacterium atrosepticum_ proteome:
```bash
< 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     124     206
< 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     243     315
< 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     3       79
> 19D654888B7BBAA5F25D864AC3EED452      PIRSF000124     4       418
< 784743901CB1C06EFA8E81D95A4E5EEF      PIRSF500136     1       375
> 9135B60EF6DD1A3FF518EEAC180C869E      PIRSF001220     1       335
> 950E62E86EB3FE4811986E4F375839C7      PIRSF000904     1       333
> E32EFA493FCA065EC24E6BB786B9621B      PIRSF000437     19      814
< EA3F94C5A9E7A263380DFEB235C07993      PIRSF037663     1       132
============ Summary ============
Matches only in expected : 5
Matches only in observed : 4
Matches in both          : 749
```
Testing using a _Dickeya dadantii_ proteome:
```bash
> 15BDC68D63495CFCFFD4662E491803B2      PIRSF000124     4       420
> 2A3CB5C20FC06DBBA1104E10FE14B0F9      PIRSF000904     1       333
< 6A8CF5C9F76BAFA035AF7523E87A0259      PIRSF000770     7       285
> 8FEAD2286E48A35C521C75EA04B6E106      PIRSF000437     19      813
> E3E1975FB197EE8884B5D74278E354B7      PIRSF016636     1       467
< E8639F7660330407EE01F92F285C15F0      PIRSF500136     1       368
> F474288DB0D5D17951191653FDB8CCEC      PIRSF001220     1       335
============ Summary ============
Matches only in expected : 2
Matches only in observed : 5
Matches in both          : 695
```

laisecf commented:
> I'm consistently finding differences when comparing the output of `PARSE_PIRSF` and the perl script for various [proteomes](https://github.com/user-attachments/files/20300958/domtbl_files.zip). It's a small number by `PARSE_PIRSF` is a groovy implementation of the perl module so we shouldn't be seeing any differences when the input is the same.
> 
> Testing using a _Phytophthora megakarya_ proteome:
> 
> ```shell
> < 1224CA081163A9B391E6EFE4CB065391      PIRSF000077     1       104
> > 294A405EB457227AC5E88CA9DFD519B8      PIRSF001220     62      406
> > 88C144D91676159D802B9271FB2464D8      PIRSF019663     9       238
> > BC3C9B7C2BFD53A9C3706A04F5B4F6EC      PIRSF000778     1       132
> < BE495BC3E429D23BB6DC48DC8C65CFB1      PIRSF000654     7       263
> > FB5A78AE553D9E05A80C77A9268D957D      PIRSF000904     7       333
> ============ Summary ============
> Matches only in expected : 2
> Matches only in observed : 4
> Matches in both          : 627
> ```
> 
> Testing using a _Fusarium oxysporum_ proteome:
> 
> ```shell
> > 0917AEB0B21ED38129863686AD88BD43      PIRSF000778     1       137
> > 2B88E803A198E201C4772AFBC3E37A60      PIRSF001220     18      73
> > 2B88E803A198E201C4772AFBC3E37A60      PIRSF001220     89      386
> > 34C7DCFF982954C7FE7ABB61956B6382      PIRSF000124     14      91
> > 34C7DCFF982954C7FE7ABB61956B6382      PIRSF000124     501     601
> > 34C7DCFF982954C7FE7ABB61956B6382      PIRSF000124     97      477
> > 3FAD2A5DA39E7C0B6E20A4748BE1B77F      PIRSF000778     14      154
> > 4C1189E51A8C227FB859BB7743596AE7      PIRSF000124     76      451
> > 7018EE411546F8029C4130D023B22C17      PIRSF019663     4       329
> < 7ABD13AC026ACDB0EC99A1684C8B4A85      PIRSF036603     463     560
> < 7ABD13AC026ACDB0EC99A1684C8B4A85      PIRSF036603     9       274
> > A23F4E183B1D1B3CC7B8FF3B2446657F      PIRSF001220     1       225
> > BCC5562BE28E9028E0CE54B5C212A631      PIRSF000124     14      28
> > BCC5562BE28E9028E0CE54B5C212A631      PIRSF000124     158     538
> > BCC5562BE28E9028E0CE54B5C212A631      PIRSF000124     562     662
> > BCC5562BE28E9028E0CE54B5C212A631      PIRSF000124     82      152
> > C2542CC2AB3292AD8DD57C01A4A48858      PIRSF000124     49      429
> > CE414C5462C4094CACA00A3001D45B81      PIRSF000124     50      423
> > E6B761F0C53F73524F8926B1BAAFDF61      PIRSF000124     63      432
> > EB8C6264D585C439A907B7CE64D3A9FC      PIRSF000904     1       250
> ============ Summary ============
> Matches only in expected : 2
> Matches only in observed : 18
> Matches in both          : 1248
> ```
> 
> Testing using a _Pectobacterium atrosepticum_ proteome:
> 
> ```shell
> < 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     124     206
> < 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     243     315
> < 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     3       79
> > 19D654888B7BBAA5F25D864AC3EED452      PIRSF000124     4       418
> < 784743901CB1C06EFA8E81D95A4E5EEF      PIRSF500136     1       375
> > 9135B60EF6DD1A3FF518EEAC180C869E      PIRSF001220     1       335
> > 950E62E86EB3FE4811986E4F375839C7      PIRSF000904     1       333
> > E32EFA493FCA065EC24E6BB786B9621B      PIRSF000437     19      814
> < EA3F94C5A9E7A263380DFEB235C07993      PIRSF037663     1       132
> ============ Summary ============
> Matches only in expected : 5
> Matches only in observed : 4
> Matches in both          : 749
> ```
> 
> Testing using a _Dickeya dadantii_ proteome:
> 
> ```shell
> > 15BDC68D63495CFCFFD4662E491803B2      PIRSF000124     4       420
> > 2A3CB5C20FC06DBBA1104E10FE14B0F9      PIRSF000904     1       333
> < 6A8CF5C9F76BAFA035AF7523E87A0259      PIRSF000770     7       285
> > 8FEAD2286E48A35C521C75EA04B6E106      PIRSF000437     19      813
> > E3E1975FB197EE8884B5D74278E354B7      PIRSF016636     1       467
> < E8639F7660330407EE01F92F285C15F0      PIRSF500136     1       368
> > F474288DB0D5D17951191653FDB8CCEC      PIRSF001220     1       335
> ============ Summary ============
> Matches only in expected : 2
> Matches only in observed : 5
> Matches in both          : 695
> ```

I think now I got the last issue for those mismatches! It was chosen superfamily (PIRSF5...) matches as the best matches, I fix the line that avoid it and all mismatches solved for Pectobacterium atrosepticum proteome. 
I am just checking some other proteomes to make sure we have no more mismatches for same hmmer input and ask again for the review ^^

HobnobMancer reviewed (changes requested):
Sorry, it's so close :cry: i6 seems to be a little too lenient for some reason - it is selecting matches that i5 is rejecting.

For a _Pectobacterium atrosepticum_:
```bash
> 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     124     206
> 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     243     315
> 14FA6749182AC5D0E9A1F4AB49879BA5      PIRSF005739     3       79
> EA3F94C5A9E7A263380DFEB235C07993      PIRSF037663     1       132
============ Summary ============
Matches only in the perl output : 0
Matches only in the i6 output   : 4
Matches in both                 : 753
```

For a _Dickeya dadantii_ genome:
```bash
> 6A8CF5C9F76BAFA035AF7523E87A0259      PIRSF000770     7       285
============ Summary ============
Matches only in the perl output : 0
Matches only in the i6 output   : 1
Matches in both                 : 700
```
For a _Fusarium oxysporum_ genome:
```bash
$ python3 pirsf/compare_pirsf.py 
> 6A8CF5C9F76BAFA035AF7523E87A0259      PIRSF000770     7       285
============ Summary ============
Matches only in the perl output : 0
Matches only in the i6 output   : 1
Matches in both                 : 700
```
For a _Phytophthora megakarya_ genome:
```bash
> 1224CA081163A9B391E6EFE4CB065391      PIRSF000077     1       104
> BE495BC3E429D23BB6DC48DC8C65CFB1      PIRSF000654     7       263
============ Summary ============
Matches only in the perl output : 0
Matches only in the i6 output   : 2
Matches in both                 : 631
```

My guess would be the initial filtering of the matches is too lenient, maybe?
With Google Drive access being limited I'll send over data files that might help over slack

laisecf commented:
The precision improvement fix some of the mismatches but now the difference of the remains (for human and Pectobacterium proteomes) are the different scores that comes from hmmsearch). E.g.:
Pectobacterium atrosepticum proteome
```
> EA3F94C5A9E7A263380DFEB235C07993   PIRSF037663  1   132

i5:
PIRSF037663
score: 38.8, minS: 39,
STORED

I6:
PIRSF037663 Protein: EA3F94C5A9E7A263380DFEB235C07993, minS: 39, Score: 39.3
ENTER ON FIRST ELSE IF
```

Those cases have a small difference but it makes difference for this block:
```
            } else if (
                r > LENGTH_RATIO_THRESHOLD &&
                    ovl >= OVERLAP_THRESHOLD &&
                    match.score >= datEntries[modelAccession].minS &&
                    (ld < LENGTH_DEVIATION_THRESHOLD * datEntries[modelAccession].stdL || ld < MINIMUM_LENGTH_DEVIATION)
            ) {
``` 
on this part:
`match.score >= datEntries[modelAccession].minS`

matthiasblum commented:
I spent some time looking at PIRSF last night, and there are a few things to change, even if this means having differences with InterProScan5:

- we need to exclude partial domains. When using the full output of hmmsearch, we can check if a domain/location is ok but checking `location.included` . If it it's `false`, then we should ignore it. This means we need to store the full hmmsearch output in `SEARCH_PIRSF` instead of the domain tab output. Since we need the length of the sequences in `PARSE_PIRSF`, we can pass the fasta file and load it with `FastaFile.parse()`
- We correctly ["merge" individual domains](https://github.com/ebi-pf-team/interproscan6/blob/8fdc7276198fdc57c5d242bb4422890b85091bf5/modules/pirsf/main.nf#L107-L110) (even if they do not overlap, the comment is wrong), to check if the family should be considered or not (based on its coverage), but we still [use the individual locations when keeping filtered match](https://github.com/ebi-pf-team/interproscan6/blob/8fdc7276198fdc57c5d242bb4422890b85091bf5/modules/pirsf/main.nf#L152). If we use the merged location to filter, we should report the merged location as well, not the original individual locations. PIRSF is a database of protein families, and we expect only on reported match/hit/location/region per PIRSF model.
- We use the [envelope start/end instead of the alignment start/end and the alignment start/end instead of the hmm start/end](https://github.com/ebi-pf-team/interproscan6/blob/8fdc7276198fdc57c5d242bb4422890b85091bf5/modules/pirsf/main.nf#L201-L204). I don't know why InterProScan 5 does this, but it doesn't make sense to me.

HobnobMancer commented:
> We use the [envelope start/end instead of the alignment start/end and the alignment start/end instead of the hmm start/end](https://github.com/ebi-pf-team/interproscan6/blob/8fdc7276198fdc57c5d242bb4422890b85091bf5/modules/pirsf/main.nf#L201-L204). I don't know why InterProScan 5 does this, but it doesn't make sense to me.

I think the might be envelope used in order to find a match with the largest sequence coverage because a protein can only be assigned to one family so you want the family to represent as much of the sequence as possible? From the [PIRSF docs](https://proteininformationresource.org/pirwww/about/doc/PIRSF.pdf):

> Proteins will no longer be constrained to belong to one and only one superfamily (but still
in one and only one homeomorphic family).

Or it is simply a bug in the [`process_results()`](https://github.com/ebi-pf-team/interproscan/blob/27c44e0c54984e6021e04f3212699098613f07c7/core/jms-implementation/support-mini-x86-32/bin/pirsf/3.10/PIRSF.pm#L152-L197) and/or [`process_hit()`](https://github.com/ebi-pf-team/interproscan/blob/27c44e0c54984e6021e04f3212699098613f07c7/core/jms-implementation/support-mini-x86-32/bin/pirsf/3.10/PIRSF.pm#L199-L235) functions in the perl script - e.g. the indexes to extract the data from dtbl file could just be wrong?

Do we have the original Georgetown script that the perl and this groovy script are based on/inspired by?

matthiasblum reviewed (changes requested):
Looks in good shape. There are a few things to address, but overall it's working as expected.

<details>
<summary>Example sequences</summary>

```
>tr|A0A009HZD2|A0A009HZD2_9GAMM Conserved region in glutamate synthase family protein OS=Acinetobacter sp. 1295259 OX=1310608 GN=J507_2520 PE=3 SV=1
MASPAQIIRHKFLNTFFSRHSLWFLCIFTALSITWFRTVYTPHSIYYFISDTLLNGLWVV
TGALSLLGLYDVLQNKHSILRNYPIMGHFRFLFEDIRPEIRQYFIEADQDALPFSRMQRS
LVYQRAKNENADKPFGSIIDVYQENYRFIVHSISPCPPADPKSFRIQVGNAQCLQPYSAS
IMNISAMSFGSLSANAIRALNKGAQLGGFYHDTGEGSLSPYHLENGGDIVWQIASGYFGC
RTLDGKFDEQKFAKQSQLPQIKMIELKLSQGAKPGHGGILPKHKITEEIAQIRGVSRDHD
CISPAKHSSFSTPIEMMHFLQKLRTLSSGKPVGFKLCIGQPWQFMSIVKAMLETKIVPDF
IVVDGSEGGTGAAPIEFSDNIGTPLREGLRFVHNTLVGAGLRDQVKIGASGKIISAFDIA
STFALGADWVNSARGFMFAVGCIQAQSCHTNQCPVGVATQDKDRQKAIDVPTKAERVFNF
HKNTLHALSEMIAAAGLRHPSDIKAHHLAQRVNDREIKNYAQLHFFLKEGELLQSELNND
DDNFYYRMWNMADANHF
>tr|A0A1B8XUR4|A0A1B8XUR4_XENTR Protein kinase domain-containing protein OS=Xenopus tropicalis OX=8364 GN=XENTR_v90026860mg PE=4 SV=1
MPNGSLAIYRVSNEDAGKYTCIAGNACDIKHRDTFLYVVDKPPGGASETDSPYKLIQTIV
LSVVAAVTYIIIVLGLMLYCKKRRKARRRDKPEGEEPEMECLNGGTLLQNGQTTAEIPEE
VPLTTLGNKRHSSGDKITFPRASLHPITTLGRGEFGEVFLAKAPSADSAAGESVVLVKAL
QTRDEQLQMDFRRELDMFSKLNHSNVVRLVGQCREAEPHYMVLEYVDLGDLKQFLRISRS
RDEKPKPLSSKHKVALGMEHLSNSRFVHKDLAARNCLVSAQRLVKVSALGLSKDVYSSEY
HPLRQAKVPLRWMPPEAVQEDDFSTKSDVWSFGVLMWEVFTLGELPYTSLPDEEVLAGLQ
NGSLKLAAPEGCSSRIYRLMQRCWAPSPKDRPSFSEIVNTLGDLTSSSKTQG
>sp|P0A9U9|YDIO_ECOL6 Probable acyl-CoA dehydrogenase YdiO OS=Escherichia coli O6:H1 (strain CFT073 / ATCC 700928 / UPEC) OX=199310 GN=ydiO PE=3 SV=1
MDFSLTEEQELLLASIRELITTNFPEEYFRTCDQNGTYPREFMRALADNGISMLGVPEEF
GGIPADYVTQMLALMEVSKCGAPAFLITNGQCIHSMRRFGSAEQLRKTAESTLETGDPAY
ALALTEPGAGSDNNSATTTYTRKNGKVYINGQKTFITGAKEYPYMLVLARDPQPKDPKKA
FTLWWVDSSKPGIKINPLHKIGWHMLSTCEVYLDNVEVEESDMVGEEGMGFLNVMYNFEM
ERLINAARSTGFAECAFEDAARYANQRIAFGKPIGHNQMIQEKLALMAIKIDNMRNMVLK
VAWQADQHQSLRTSAALAKLYCARTAMEVIDDAIQIMGGLGYTDEARVSRFWRDVRCERI
GGGTDEIMIYVAGRQILKDYQNK
>tr|Q57YS8|Q57YS8_TRYB2 Protein kinase, putative OS=Trypanosoma brucei brucei (strain 927/4 GUTat10.1) OX=185431 GN=Tb08.10K10.520 PE=4 SV=1
MKNNDANHMRKRSREDPTNAEAAEFSEDQGVGGQQTEVKGTKSSSTSRVSPEIPDESACI
YEYVSGPTAVSTSGGNSGNSPGSTHPSERKLKQATLPSYGLVNDTAVFRKELADRDAQID
ELREKLAAVETRMSERETTLVNTQAQLQEAVDRNNRYQVVLREEMLRAARQERCDARRAL
HLKQFELGQIAVWHSNGREVWVEGNKMRQLTMQLEELSVRRDEVEELKKTAEKRARQILR
SNDEDSMAPEAQAALMELQEAALLYTAEFAALGSAIQSLKQQQQDLNHEKKVFLKEIRRV
NDEDASAFVAVPALGHNGRYVMMHLLGKGGFSEVWKAFDLQEARYVACKIHRVQREWSQQ
VRQHYRDRAVRELKIMRMLEHPHLTRLFDAFDHGTATFVSVMEFSAGTDLDTHLKRCGTL
REVEARLIIMQVVSALRYFAAQHQPVIHYDLKPANILLHSSNQSSLLIKITDFGLSKLIP
KRDGASDNPTIELTSQGAGTYWYLPPECFDTTATPRISNKVDVWSCGVIFYQMLFGRRPF
AEGESQQQIWQNKLIVSSAHTLTFPDTPRVSQEAKDLIQKCLEYHPADRYDVMQLSQDPY
LQRNTRRLSRTERTLPAATQNSSTTVVTTAAAASGVTLTGSVEEKLSNIP
```
</details>

**InterProScan 5**

| UniProtKB | PIRSF | Start | End |
|---------------|-----------|--------|-------|
|sp\|P0A9U9\|YDIO_ECOL6|	PIRSF016578|	8|	96|
|sp\|P0A9U9\|YDIO_ECOL6|	PIRSF016578|	235|	381|
|sp\|P0A9U9\|YDIO_ECOL6|	PIRSF016578|	134|	190|
|tr\|A0A009HZD2\|A0A009HZD2_9GAMM|	PIRSF006429|	32|	557|
|tr\|A0A009HZD2\|A0A009HZD2_9GAMM|	PIRSF500060|	35|	557|
|tr\|A0A1B8XUR4\|A0A1B8XUR4_XENTR|	PIRSF000615|	2|	43|
|tr\|A0A1B8XUR4\|A0A1B8XUR4_XENTR|	PIRSF000615|	245|	412|
|tr\|A0A1B8XUR4\|A0A1B8XUR4_XENTR|	PIRSF000615|	134|	251|
|tr\|A0A1B8XUR4\|A0A1B8XUR4_XENTR|	PIRSF000615|	44|	87|
|tr\|Q57YS8\|Q57YS8_TRYB2|	PIRSF000654|	3|	56|
|tr\|Q57YS8\|Q57YS8_TRYB2|	PIRSF000654|	341|	597|

**InterProScan 6**

| UniProtKB | PIRSF | Start | End |
|---------------|-----------|--------|-------|
|sp\|P0A9U9\|YDIO_ECOL6|	PIRSF016578|	31|	357|
|tr\|A0A009HZD2\|A0A009HZD2_9GAMM|	PIRSF006429|	56|	553|
|tr\|A0A009HZD2\|A0A009HZD2_9GAMM|	PIRSF500060|	48|	557|
|tr\|A0A1B8XUR4\|A0A1B8XUR4_XENTR|	PIRSF000615|	5|	410|

InterProScan 6 is losing the match in `Q57YS8`, so I want to check this sequence, but it might correct.Q57YS8

laisecf commented on modules/pirsf/main.nf:
```diff
@@ -94,80 +50,83 @@ process PARSE_PIRSF {
+            double ovl = Math.abs(seqEnd - seqStart + 1) / sequenceLength
             // Ratio over coverage of sequence and profile hmm
-            double r = Math.abs(match.locations[0].hmmEnd - match.locations[0].hmmStart + 1) / (match.locations[0].envelopeEnd - match.locations[0].envelopeStart + 1)
+            double r = Math.abs(hmmEnd - hmmStart + 1) / (seqEnd - seqStart + 1)
```
Why to remove the `+ 1`?

laisecf commented on modules/pirsf/main.nf:
```diff
@@ -94,80 +50,83 @@ process PARSE_PIRSF {
             if (datChildren.containsKey(modelAccession)) {
                 // process a subfamily match
-                if (r > LENGTH_RATIO_THRESHOLD && match.locations[0].score >= datEntries[modelAccession].minS) {
+                if (r > LENGTH_RATIO_THRESHOLD && match.score >= datEntries[modelAccession].minS) {
```
The same for the final score when building the merged location, right?
